runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -135,31 +149,131 @@ public void open() throws Exception {
+
+    private final StateInternals stateInternals;
+    private final KeyedStateBackend<ByteBuffer> keyedStateBackend;
+
+    private BagUserStateFactory(
+        StateInternals stateInternals, KeyedStateBackend<ByteBuffer> keyedStateBackend) {
+
+      this.stateInternals = stateInternals;
+      this.keyedStateBackend = keyedStateBackend;
+    }
+
+    @Override
+    public <K, V, W extends BoundedWindow>
+        StateRequestHandlers.BagUserStateHandler<K, V, W> forUserState(
+            String pTransformId,
+            String userStateId,
+            Coder<K> keyCoder,
+            Coder<V> valueCoder,
+            Coder<W> windowCoder) {
+      return new StateRequestHandlers.BagUserStateHandler<K, V, W>() {
+        @Override
+        public Iterable<V> get(K key, W window) {
+          prepareStateBackend(key, keyCoder);
+          StateNamespace namespace = StateNamespaces.window(windowCoder, window);
+          BagState<V> bagState =
+              stateInternals.state(namespace, StateTags.bag(userStateId, valueCoder));
+          return bagState.read();
+        }
+
+        @Override
+        public void append(K key, W window, Iterator<V> values) {
+          prepareStateBackend(key, keyCoder);
+          StateNamespace namespace = StateNamespaces.window(windowCoder, window);
+          BagState<V> bagState =
+              stateInternals.state(namespace, StateTags.bag(userStateId, valueCoder));
+          while (values.hasNext()) {
+            bagState.add(values.next());
+          }
+        }
+
+        @Override
+        public void clear(K key, W window) {
+          prepareStateBackend(key, keyCoder);
+          StateNamespace namespace = StateNamespaces.window(windowCoder, window);
+          BagState<V> bagState =
+              stateInternals.state(namespace, StateTags.bag(userStateId, valueCoder));
+          bagState.clear();
+        }
+
+        private void prepareStateBackend(K key, Coder<K> keyCoder) {
+          ByteArrayOutputStream baos = new ByteArrayOutputStream();
+          try {
+            keyCoder.encode(key, baos);
+          } catch (IOException e) {
+            throw new RuntimeException("Failed to encode key for Flink state backend", e);
+          }
+          keyedStateBackend.setCurrentKey(ByteBuffer.wrap(baos.toByteArray()));
+        }
+      };
     }
   }
 
+  @Override
+  public void setKeyContextElement1(StreamRecord record) throws Exception {
+    // Note: This is only relevant when we have a stateful DoFn.
+    // We want to control the key of the state backend ourselves and
+    // we must avoid any concurrent setting of the current active key.
+    // By overwriting this, we also prevent unnecessary serialization
+    // as the key has to be encoded as a byte array.
+  }
+
+  @Override
+  public void setCurrentKey(Object key) {
+    throw new IllegalStateException(
+        "This should never be called. Current key for StateBackend should be set manually.");
```

Comment:
Maybe use UnsupportedOperationException and "Current key for state backend can only be set by SDK workers."